### PR TITLE
[WIP] Add support for updating the on-chain multisig safe address on deployment

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR add support for fetching and updating the multisig safe address when deploying from tokens 

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [x] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

### CODE REVIEW

File: autonomy/cli/deploy.py [hunk 1]
```diff
@@ -272,6 +280,26 @@ def run_deployment_from_token(  # pylint: disable=too-many-arguments, too-many-l
...
+    for override in overrides:
+        if override["type"] == SKILL:
+            addresses = override["setup_args"]["args"]["setup"]["safe_contract_address"]
+            override["setup_args"]["args"]["setup"]["safe_contract_address"] = [
+                *addresses,
+                address,
+            ]
+
+    with open_file(service_path / DEFAULT_SERVICE_FILE, mode="w+") as fp:
+        yaml_dump_all([config, *overrides], fp)
```
Comment (angrybayblade) [inline autonomy/cli/deploy.py:300]: This was not supposed to be part of this PR, I'll revert this and create a proper PR later. This is just here for the context

File: autonomy/deploy/chain.py [hunk 2]
```diff
@@ -105,6 +107,24 @@ def resolve_token_id(
...
+        Returns service info.
+
+        :param token_id: Token ID pointing to the on-chain service
+        :returns: security deposit, multisig address, IPFS hash for config,
+                threshold, max number of agent instances, number of agent instances,
+                service state, list of cannonical agents
+        """
+
+        info = self.contract.functions.getService(token_id).call()
+        return info
```
Comment (angrybayblade) [inline autonomy/deploy/chain.py:130]: To retrieve service info

File: autonomy/test_tools/docker/registries.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+COMPONENT_REGISTRY = "0x5FbDB2315678afecb367f032d93F642f64180aa3"
+AGENT_REGISTRY = "0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512"
+REGISTRIE_SMANAGER = "0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0"
+GNOSIS_SAFE_MULTISIG = "0xA51c1fc2f0D1a1b8494Ed1FE312d7C3a78Ed91C0"
+SERVICE_REGISTRY = "0x0DCd1Bf9A1b36cE34237eEaFef220932846BCD82"
+SERVICE_MANAGER = "0x9A676e781A523b5d0C0e43731313A708CB607508"
+SERVICE_MULTISIG = "0xe1bB37cb3Dd284B490C874C1F3f414FbbE2C278e"
+
+
+class RegistriesDockerImage(DockerImage):
```
Comment (angrybayblade) [inline autonomy/test_tools/docker/registries.py:49]: Registries test image

File: packages/valory/services/oracle_hardhat/service.yaml [hunk 4]
```diff
@@ -10,15 +10,17 @@ fingerprint_ignore_patterns: []
...
 number_of_agents: 4
 ---
-benchmark_persistence_params:
-  args: &id003
-    log_dir: /benchmarks
-observation_params:
+setup_args:
   args: &id001
+    setup:
+      safe_contract_address: []
```
Comment (angrybayblade) [inline packages/valory/services/oracle_hardhat/service.yaml:16]: To support service level override for safe address

File: scripts/run_service_registry_locally.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+@click.command()
+@click.argument(
+    "third_party_contracts",
+    type=click.Path(),
+    required=False,
+    default=DEFAULT_THIRD_PARTY_PATH,
+)
+def main(third_party_contracts: Path) -> None:
+    """Run a local registry deployment."""
```
Comment (angrybayblade) [inline scripts/run_service_registry_locally.py:40]: This is just a utility script to run a registry deployment for development

File: autonomy/cli/deploy.py [hunk 6]
```diff
@@ -272,6 +280,26 @@ def run_deployment_from_token(  # pylint: disable=too-many-arguments, too-many-l
...
+        config, *overrides = yaml_load_all(
+            fp,
+        )
+
+    for override in overrides:
+        if override["type"] == SKILL:
+            addresses = override["setup_args"]["args"]["setup"]["safe_contract_address"]
+            override["setup_args"]["args"]["setup"]["safe_contract_address"] = [
+                *addresses,
+                address,
```
Comment (DavidMinarsch) [inline autonomy/cli/deploy.py:296]: There's only one possible address, so I would not append the list.

Comment (angrybayblade) [reply autonomy/cli/deploy.py:296]: So why keep it as the list? can't we declare it as a string?

File: autonomy/cli/deploy.py [hunk 7]
```diff
@@ -225,7 +228,7 @@ def run_deployment_from_token(  # pylint: disable=too-many-arguments, too-many-l
     """Run service deployment."""
 
     ctx = cast(Context, click_context.obj)
-    ctx.registry_type = registry
+    ctx.registry_type = REGISTRY_REMOTE
```
Comment (angrybayblade) [inline autonomy/cli/deploy.py:231]: We use remote registry as default to fetch packages when deploying using token

File: README.md [hunk 8]
```diff
@@ -51,6 +51,11 @@ Ensure your machine satisfies the following requirements:
       cd ../..
       cd third_party/contracts-amm && yarn install
       cd ../..
+      cd thid_party/autonolas-registries
+      git submodule init
+      git submodule sync --recursive
+      git submodule update --recursive
```
Comment (DavidMinarsch) [inline README.md:57]: This is already described on line 46/42

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]
Note, your Windows CI setup is not correct

File: scripts/run_service_registry_locally.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    client = from_env()
+    image = RegistriesDockerImage(
+        client=client, third_party_contract_dir=third_party_contracts
+    )
+    container = image.create()
+
+    try:
+        for line in client.api.logs(container.id, follow=True, stream=True):
+            print(line.decode(), end="")
+    except KeyboardInterrupt:
```
Comment (Karrenbelt) [inline scripts/run_service_registry_locally.py:54]: Why separate catch?

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**